sklearn/tree/export.py

```diff
@@ -397,6 +398,23 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
             return_string = True
             out_file = six.StringIO()
 
+        # Check length of feature_names before get into the tree node
+        # raise error for too few feature_names
+        if feature_names is not None:
+            if len(feature_names) < decision_tree.n_features_:
+                raise ValueError("Length of feature_names=%d "
+                                 "does not match number of features=%d"
+                                 % (len(feature_names),
+                                    decision_tree.n_features_))
+            # for too much feature_names, will use the first n_features
+            # raise an warning for users
+            if len(feature_names) > decision_tree.n_features_:
```

Comment:
Not sure that this is appropriate, I would just raise an error they are not equal. It is dangerous to guess that just selecting the first n values of feature_names is what the user intended.